### PR TITLE
clock_control: nrf5_power: Let the LF clock be configured at least once

### PR DESCRIPTION
**Pull request by durub**
If the LF clock was already started, but wasn't initialized with
_k32src_start yet, allow it to run once. This is needed because if a
soft reset is triggered while watchdog is active, the LF clock will
already be running, but won't be configured yet (watchdog forces LF
clock to be running).

If the LF clock isn't configured with _k32src_start, the nRF5 RTC
won't count and the whole system will malfunction, as the kernel
depends on the clock source being working properly.

Signed-off-by: Thiago Silveira <thiago@exati.com.br>

### CODE REVIEW

**Comment by durub** [comment]
More information here: https://lists.zephyrproject.org/pipermail/zephyr-users/2017-September/000725.html

**Comment by durub** [comment]
To test, try this code before and after the patch:

```
#include <zephyr.h>
#include <misc/printk.h>
#include <nrf_wdt.h>

void main(void)
{
    printk("Hello World! %s\n", CONFIG_ARCH);    

    printk("Hardware cycle count: %d\n", k_cycle_get_32());
    printk("Hardware cycle count: %d\n", k_cycle_get_32());
    printk("Hardware cycle count: %d\n", k_cycle_get_32());

    printk("Before busywait: %d\n", k_uptime_get_32());
    k_busy_wait(1000000);
    printk("After busywait: %d\n", k_uptime_get_32());

    nrf_wdt_reload_request_disable(NRF_WDT_RR1);
    nrf_wdt_reload_request_disable(NRF_WDT_RR2);
    nrf_wdt_reload_request_disable(NRF_WDT_RR3);
    nrf_wdt_reload_request_disable(NRF_WDT_RR4);
    nrf_wdt_reload_request_disable(NRF_WDT_RR5);
    nrf_wdt_reload_request_disable(NRF_WDT_RR6);
    nrf_wdt_reload_request_disable(NRF_WDT_RR7);

    nrf_wdt_reload_request_enable(NRF_WDT_RR0);

    nrf_wdt_reload_request_set(NRF_WDT_RR0);
    nrf_wdt_behaviour_set(0x09);
    nrf_wdt_reload_value_set(10000 * 32678 / 1000);
    nrf_wdt_int_disable(NRF_WDT_INT_TIMEOUT_MASK);
    nrf_wdt_event_clear(NRF_WDT_EVENT_TIMEOUT);
    nrf_wdt_task_trigger(NRF_WDT_TASK_START);
}

```

It should fail after every soft reset (a reset using nrfjprog, for example). After some time, it will restart and clear the registers, letting it work again. However, it keeps failing after each and every soft reset.

File: drivers/clock_control/nrf5_power_clock.c [hunk 1]
```diff
@@ -141,15 +143,35 @@ static int _k32src_start(struct device *dev, clock_control_subsys_t sub_system)
...
 	 */
+	imask = irq_lock();
 
-	if ((NRF_CLOCK->LFCLKSTAT & CLOCK_LFCLKSTAT_STATE_Msk)) {
+	if (k32src_guard) {
+		irq_unlock(imask);
+		goto lf_already_started;
+	}
+
+	if (k32src_initialized_once &&
```
Comment (cvinayak) [inline drivers/clock_control/nrf5_power_clock.c]: if already initialized once then irq_unlock  and goto lf_already_started. Refer to line 35 to be consistent. Check for  h/w stat is not needed if the design in _m16src_start is followed.

**Review by cvinayak — CHANGES_REQUESTED** [review]

File: drivers/clock_control/nrf5_power_clock.c (hunk 1 above)
Comment (durub) [reply drivers/clock_control/nrf5_power_clock.c]: I was doing that at first for consistency, but I changed it later for a little more correctness. Doesn't that add a second `if` (equal to the `goto` check) which, if pre-empted away with a _just right_ timing, may return a different value, returning `-EINPROGRESS` instead of 0? That may be too minor an issue to worry about though, even if I would prefer it to be bulletproof...

About the h/w stat check, I think it's needed here. The design used in `_m16src_start` doesn't apply here, because we have no `stop` function implemented. Because of this, there's no need for a ref count -- it would never decrease. If we want to disallow callers to call it more than once, that's fine. My intention was to allow the caller to call the function at least once (even if the peripheral is already on, to configure it) and, after this, allow the caller to call it every time the peripheral was disabled. The use case for this is if the user manually stops it outside of the driver for some reason. What do you think?

Comment (cvinayak) [reply drivers/clock_control/nrf5_power_clock.c]: _m16src_start enters with irq_lock, hence I dont see what gets pre-empted away. If _m16src_start is not bulletproof it needs fixing.

(k32src_initialized_once == 1) implies (NRF_CLOCK->LFCLKSTAT & CLOCK_LFCLKSTAT_STATE_Msk) != 0. Else something is wrong in following code that the code has not started the LF clock after the first time k32src_initialized_once  was set. Do you agree?

**Review by cvinayak — CHANGES_REQUESTED** [review]

File: drivers/clock_control/nrf5_power_clock.c (hunk 1 above)
Comment (durub) [reply drivers/clock_control/nrf5_power_clock.c]: @cvinayak, `_m16src_start` is fine. From what I can see, the ref count handles all cases correctly.

About the 32 KHz, looking from the point of view of the driver, your assessment is correct and I agree. What I meant is that, if the user, for some reason, wanted to control the peripheral directly (i.e `NRF_CLOCK->TASKS_LFCLKSTOP = 1`), the driver wouldn't work as they expected anymore. Example:

```
struct device *clock;

clock = device_get_binding(CONFIG_CLOCK_CONTROL_NRF5_K32SRC_DRV_NAME);
if (!clock) {
    return -1;
}

clock_control_on(clock, (void *) CLOCK_CONTROL_NRF5_K32SRC);

NRF_CLOCK->TASKS_LFCLKSTOP = 1;

/* wait until clock stops */

/* wouldn't work as expected by the user; the 32 KHz clock source wouldn't 
 * turn on/be configured again
 */
clock_control_on(clock, (void *)CLOCK_CONTROL_NRF5_K32SRC);
```

I do concede that this is a very niche use case, but I like when things don't fail in unexpected conditions. However, seeing as the high frequency clock source doesn't support this, there seems to be no need to support it in the low frequency clock source. Consistency is king.

Then, do you agree with the following course of action?

1) Remove the guard, as it is not needed anymore. The initialized once check already handles this case well;
2) Remove the h/w stat check, keeping only the k32src_initialized_once check. If it fails, goto already started.

An alternative course of action would be to implement the stop function and the ref count. If you think that is better, I would gladly do it. My initial focus was only on solving the bug, but I could expand this PR to include the stop feature.

Thanks for the attention,
Thiago

Comment (cvinayak) [reply drivers/clock_control/nrf5_power_clock.c]: Hi Thiago,

I agree to your 2 step resolution. This is the simplest.

Lets not add ref count and stop implementations, as it will not be used yet (as it was before). I agree we solve the bug.

Thanks and regards,
Vinayak

File: drivers/timer/nrf_rtc_timer.c [hunk 2]
```diff
@@ -11,7 +11,6 @@
 #include <arch/arm/cortex_m/cmsis.h>
 #include <sys_clock.h>
 
-extern void youve_print(void);
```
Comment (cvinayak) [inline drivers/timer/nrf_rtc_timer.c]: Exclude this file  change from this commit (split into another commit/PR with appropriate commit message)

**Review by cvinayak — CHANGES_REQUESTED** [review]

File: drivers/clock_control/nrf5_power_clock.c (hunk 1 above)
Comment (durub) [reply drivers/clock_control/nrf5_power_clock.c]: Done!

File: drivers/timer/nrf_rtc_timer.c (hunk 2 above)
Comment (durub) [reply drivers/timer/nrf_rtc_timer.c]: Done. I will create a PR shortly with these clean up changes.

**Review by cvinayak — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
@nashif I think this is ready for merging
